src/settings/settings_manager.cpp

```diff
@@ -0,0 +1,209 @@
+#include <gflags/gflags.h>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "common/macros.h"
+#include "main/db_main.h"
+#include "settings/settings_manager.h"
+#include "type/transient_value_factory.h"
+
+#define __SETTING_GFLAGS_DECLARE__     // NOLINT
+#include "settings/settings_common.h"  // NOLINT
+#include "settings/settings_defs.h"    // NOLINT
+#undef __SETTING_GFLAGS_DECLARE__      // NOLINT
+
+namespace terrier::settings {
+
+using ValueFactory = type::TransientValueFactory;
+using ValuePeeker = type::TransientValuePeeker;
+using ActionContext = common::ActionContext;
+using ActionState = common::ActionState;
+
+// Used for building temporary transactions
+void EmptyCallback(void * /*unused*/) {}
+
+SettingsManager::SettingsManager(DBMain *db, catalog::Catalog *catalog)
+    : db_(db), settings_handle_(catalog->GetSettingsHandle()) {
+  ValidateParams();
+}
+
+void SettingsManager::ValidateParams() {
+  // This will expand to invoke settings_manager::DefineSetting on
+  // all of the settings defined in settings.h.
+  // Example:
+  //   ValidateSetting(Param::port, type::TransientValueFactory::GetInteger(1024),
+  //                  type::TransientValueFactory::GetInteger(65535));
+
+#define __SETTING_VALIDATE__           // NOLINT
+#include "settings/settings_common.h"  // NOLINT
+#include "settings/settings_defs.h"    // NOLINT
+#undef __SETTING_VALIDATE__            // NOLINT
+}
+
+void SettingsManager::ValidateSetting(Param param, const type::TransientValue &min_value,
+                                      const type::TransientValue &max_value) {
+  const ParamInfo &info = db_->param_map_.find(param)->second;
+  if (!ValidateValue(info.value_, min_value, max_value)) {
+    SETTINGS_LOG_ERROR(
+        "Value given for \"{}"
+        "\" is not in its min-max bounds",
+        info.name_);
+    throw SETTINGS_EXCEPTION("Invalid setting value");
+  }
+}
+
+int32_t SettingsManager::GetInt(Param param) {
+  common::SharedLatch::ScopedExclusiveLatch guard(&latch_);
```

Comment:
These don't need to be exclusive do they? SharedLatch should suffice since it's only reading.